rfcs/0024-hard-forks.md

```diff
@@ -0,0 +1,455 @@
+   S_n+3, and a merge proof using the previous one with statement S_n -> S_n+3.
+   This process continues until the scan state is empty and a snarked ledger
+   proof up to S_n is emitted. The final block of the orange phase applies the
+   sequentially built snark, giving us a valid era n block with a ledger that
+   includes all transactions up to now.
+4. GREEN: we may now create an era n+1 block, with an empty (new-style) scan
+   state, starting from the correct state hash. Coinbase amounts go back to
+   normal.
+
+#### Checking blockchain SNARKs from prior eras
+
+The part of the blockchain SNARK that verifies the previous blockchain SNARK
+will choose which verification key to use depending on the global slot of the
+previous block, along with whether era end finalization has finished. This will
+require adding fields to `Blockchain_state` and `Consensus_state`. A circuit for
+era n only references the verification keys for era n and era n-1, it can't
+check older SNARKs.
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+This will be a lot of work to implement, and complicates many pieces of Coda.
+We'll have to write some serialization logic manually, type hackery is always a
+mixed blessing, we'll add another way to add ledger proofs, and in general it
+enlarges the attack and bug surface. But we'll pay complexity costs for hard
+forks at some point and it'll be much nicer to do it ahead of time.
+
+## Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+* Why is this design the best in the space of possible designs?
+  This design is quite conservative in process design, and pays a pretty large
+  complexity cost up front to simplify future work. Cryptocurrency is not a
+  space where we can "move fast and break things", so the conservative end of
+  the spectrum is where we should be aiming. The complexity cost is justified by
+  simplifying transitions and operation within eras.
+* What other designs have been considered and what is the rationale for not choosing them?
+  * We could not have a fixed schedule, and simply do protocol upgrades "when
+    they're done". I like the regular schedule because it's a strong incentive
+    to make smaller changes, and ensure they actually get shipped regularly.
+  * Rather than doing the fancy type level feature flags, we could put booleans
+    and optional fields into the data structures. This would reduce the amount
+    of time we spend fighting the type checker, and reduce the knowledge burden
+    to work on the codebase, but the type level flags make the code more
+    readable and more reliable.
+  * We could not have the parallel scan drain and restart process, instead
+    simply keeping both versions as part of the block data, only removing from
+    the old one and only adding to the new one. This is the more obvious method,
+    but it's more complex in the steady state, and easier to get wrong. Fully
```

Comment:
It should be safe with the versioning right? We'd have two states with two different versions. Add work to the one with previous version and new transactions to the current version. There's one limitation that I can think of rn: If the tree size reduces in the new era then the number of snarks that needs to be done per block would be more than what it is with a smaller tree size. Worst case there could be (number-of-leaves-in-the-old-tree/2) fee transfers that gets added to the new tree and so the new tree should be able to accommodate that many transactions per block.